src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -501,9 +530,35 @@ private bool IsFirstTokenOnLine(TokenData tokenData1, TokenData tokenData2)
             return this.GetTriviaData(tokenData1, tokenData2).SecondTokenIsFirstTokenOnLine;
         }
 
+        /// <summary>
+        /// An approximation for the length of a token in this stream
+        /// </summary>
+        float AverageTokenLength { get; }
+        /// <summary>
+        /// Gets approximated index bounds for a token with specified position, which will definitely contain
+        /// the token, if it is present in the stream.
+        /// </summary>
+        private void GetTokenIndexBounds(int position, out int lowerBound, out int upperBound)
+        {
+            float avgLength = this.AverageTokenLength;
+            Debug.Assert(avgLength > 0);
+            lowerBound = upperBound = Bounded((int)((position - _tokens[0].Position) / avgLength), min: 0, max: _tokens.Length - 1);
+
+            while (_tokens[lowerBound].Position >= position && lowerBound > 0)
+            {
```

Comment:
this is nice, so every access cost now become from multiple binary compare to just 2-3 compare? since there will be a lot of token access, reduce this would have big impact. nice.

Reply:
indeed!  i like this a lot!